### PR TITLE
Treasure island

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] fix extra rmdir mkdir steps (comment bellow)
- [x] extract FSTree
- [x]  cleanup
- [x] confirm working as expected (yup, make funnel basically free on rebuilds for the stress app)
- [x] fix file -> folder & folder => file transitions (failing test just added)
- [x] add update support (for those poor souls without symlinks, but it will unlock the ability to use fs-tree for broccoli-filter) as well.

``` js
tree.reduce(preOrderReducer, postOrderReducer, [])
```

piratebroccoli

### CODE REVIEW

File: tests/fs-tree-test.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+          expect(fsTree.calculatePatch([
+            'bar/three.js'
+          ])).to.deep.equal([
+            ['unlink', 'foo/one.js'],
+            ['unlink', 'foo/two.js'],
+            ['unlink', 'bar/one.js'],
+            ['unlink', 'bar/two.js'],
+            ['rmdir', 'foo'],
+            ['rmdir', 'bar'],
+            ['mkdir', 'bar'],
```
Comment (stefanpenner) [inline tests/fs-tree-test.js]: this rmdir mkdir pair is unneeded and wasteful

**Comment by stefanpenner** [comment]
i noticed:

```
[ [ 'mkdir', '' ],
  [ 'mkdir', 'assets' ],
  [ 'mkdir', 'assets/' ],
  [ 'create', 'assets//app.css' ] ]
```

is possible, which means `assets` and `asserts/` are identified differently.

**Comment by stefanpenner** [comment]
Also just realized. When in `copy` mode (platforms without symlinks), we will actually need to re-copy if the file changed.

Although I wanted to release this without having to add support for `updates` I suspect we have no choice. That being said, on platforms with symlinks, we can likely opt out of the potentially costly extra stats for funnel. Since stable symlinks should be sufficient.

dealing with `updates` was the plan anyways for usage within broccoli-filter

File: lib/fs-tree/entries.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
+'use strict';
+
+function Entries(entries) {
```
Comment (stefanpenner) [inline lib/fs-tree/entries.js]: this should likely just be expressed as an explicit object, rather then implicitly with closures.

File: lib/fs-tree/entries.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+      return _entry.mtime !== entry.mtime || _entry.size !== entry.size || _entry.mode !== entry.mode;
+
+    });
+  }
+
+  function diffAdditions(newEntries) {
+    var paths = entries.map(byRelativePath);
+    return newEntries.filter(function(entry) {
+      return paths.indexOf(entry.relativePath) === -1;
```
Comment (stefanpenner) [inline lib/fs-tree/entries.js]: hmm, it seems like the internal structure may be better represented by FastOrderedSet, if we are apparently performing set operations on the content.

**Comment by stefanpenner** [comment]
its alive!

**Comment by hjdivad** [comment]
@stefanpenner :metal: :heart_eyes_cat: :metal:

**Pull request merged**